### PR TITLE
Added validation logic to parser

### PR DESCRIPTION
**Pull request by joelebwf**
The parser is now ready to support validation exceptions thrown by data_model in conjunction with validator.  It places all exceptions found into a list and raises them at the end of processing (it previously raised the first exception and did not complete processing.

### CODE REVIEW

File: oblib/parser.py [hunk 1]
```diff
@@ -124,64 +155,126 @@ def from_JSON_string(self, json_string, entrypoint_name=None):
 
         json_string (str): String containing JSON
         entrypoint_name (str): Optional name of the entrypoint.
-
-        TODO: Currently this method works in some cases but testing is incomplete
         """
 
+        # Create a validation error which can be used to maintain a list of error messages
+        validation_errors = ValidationErrors("Error(s) found in input JSON")
```
Comment (cwhanse) [inline oblib/parser.py]: I think this counterproductive, `validation_errors` will never be empty and thus will always be raised

File: oblib/parser.py [hunk 2]
```diff
@@ -279,22 +387,42 @@ def from_XML_string(self, xml_string, entrypoint_name=None):
...
+                        except Exception as e:
+                            validation_errors.append(ValidationError(str(e)))
+                    else:
+                        validation_errors.append("referenced context is missing")
+                else:
+                    validation_errors.append("Element is missing a context")
 
-                entrypoint.set(tag, child.text, **kwargs)
+        # Raise the errors if necessary
+        if len(validation_errors.get_errors()) > 0:
```
Comment (cwhanse) [inline oblib/parser.py]: `if validation_errors.get_errors():` is enough, no need for `len`

File: oblib/parser.py [hunk 3]
```diff
@@ -279,22 +387,42 @@ def from_XML_string(self, xml_string, entrypoint_name=None):
...
+                        tag = child.tag
+                        tag = tag.replace("{http://xbrl.us/Solar/v1.2/2018-03-31/solar}", "solar:")
+                        tag = tag.replace("{http://fasb.org/us-gaap/2017-01-31}", "us-gaap:")
+                        tag = tag.replace("{http://xbrl.sec.gov/dei/2014-01-31}", "dei:")
+                        try:
+                            entrypoint.set(tag, child.text, **kwargs)
+                        except Exception as e:
+                            validation_errors.append(ValidationError(str(e)))
+                    else:
+                        validation_errors.append("referenced context is missing")
```
Comment (cwhanse) [inline oblib/parser.py]: different type being appended here (str instead of `ValidationError`)

File: oblib/parser.py (hunk 1 above)
Comment (joelebwf) [reply oblib/parser.py]: Its not always raised.  it only is raised if errors are actually added

File: oblib/parser.py (hunk 2 above)
Comment (joelebwf) [reply oblib/parser.py]: Will fix

File: oblib/parser.py (hunk 1 above)
Comment (cwhanse) [reply oblib/parser.py]: the if condition is len(validation_errors) > 0, and we initialize validation_errors with this message, so we always hit the raise statement

Comment (joelebwf) [reply oblib/parser.py]: You have to call append() at least once in order to add a ValidationError to ValidaitonErrors and cause the raise condition to occur.

File: oblib/parser.py (hunk 3 above)
Comment (joelebwf) [reply oblib/parser.py]: I have (a) converted everything to strings and updated the append() method to automatically convert strings to ValidationError as needed.  Note that I committed twice so wait 15 minutes and then review.

File: oblib/parser.py (hunk 1 above)
Comment (cwhanse) [reply oblib/parser.py]: You are correct, I see that.  Thanks for the patience.

**Pull request merged**